---
title: "CD33 CAR NK: D2"
author: "Ahmad Al Ajami (AG Imkeller | KGU)"
date: "`r Sys.Date()`"
output:
  html_document:
    keep_md: false
    toc: true
    toc_float: true
    toc_depth: 5
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, include=TRUE, message=FALSE, cache = FALSE}
suppressPackageStartupMessages({
library(tidyverse)
library(scran)
library(scater)
library(patchwork)
library(scDblFinder)
library(cowplot)
library(pheatmap)
library(edgeR)
library(clusterProfiler)
library(msigdbr)
library(gplots)
library(ggrepel)
library(ggplotify)
})
```

# Loading Data

```{r data, message=FALSE}
data.dir <- "Files_already_generated_with_Sevenbridges/D2/"

data <- read.csv(paste(data.dir, "Combined_CARNK_Sample2_RSEC_MolsPerCell.csv", sep = ""), skip = 7)
data$Cell_Index <- paste(data$Cell_Index, "_D2", sep = "")
```

The metadata contains Sample_Tag (nucleotide labeled antibody used to distinguish between samples), Sample_Name (patient id), multiplet status assigned by the SevenBridges pipeline

```{r metadata, message=FALSE}
# donor 2
## table with cell information (sample barcode, donor, ...)
coldata <- read.csv(paste(data.dir, "CARNK_Sample2_Sample_Tag_Calls.csv", sep = ""), skip = 7) %>%
  dplyr::mutate(invalid = (Sample_Name == "Multiplet" | Sample_Name == "Undetermined")) %>%
  dplyr::mutate(donor = "D2")
coldata$Cell_Index <- paste(coldata$Cell_Index, "_D2", sep = "")
```

## SevenBridges Output

SevenBridges pre-processing pipeline report:

```{r sb_output, message=FALSE}
sb <- data.frame(Donor = "D2",
                 Number.of.Cells = dim(data)[1],
                 Number.of.Features = dim(data[, -1])[2],
                 Median.Number.of.Features = median(rowSums(data > 0)),
                 stringsAsFactors = F)

DT::datatable(sb, 
              class = 'compact stripe hower',
              options = list(
                dom = "Bfrtip",
                scrollX = TRUE,
                paging = FALSE,
                searching = FALSE,
                info = FALSE,
                ordering = FALSE,
                columnDefs = list(list(className = 'dt-center', targets = 0:3))), rownames = FALSE)
```

## Sample IDs

Number of cells in each sample:

```{r sample_ids, message=FALSE}
table(coldata$Sample_Name)
```

We define a new column containing the group of samples:

- BeforeCoc = Before
- AfterCoc = After
- NA = Multiplets + Undetermined

```{r group_ids, message=FALSE, fig.width=10, fig.height=4}
coldata <- coldata %>%
  dplyr::mutate(group = ifelse(grepl("before", Sample_Name), "BeforeCoc",
    ifelse(grepl("after", Sample_Name), "AfterCoc", NA)))

p1 <- coldata %>%
  ggplot(aes(x = donor, fill = group)) + 
  geom_bar() +
  labs(x = "Donor", fill = "Group")

p2 <- coldata %>%
  ggplot(aes(x = donor, fill = Sample_Name)) + 
  geom_bar() +
  labs(x = "Donor", fill = "Sample Name")

p3 <- coldata %>%
  ggplot(aes(x = group, fill = Sample_Name)) + 
  geom_bar() +
  labs(x = "Group", fill = "Sample Name")

p1 + p2 + p3
```

```{r}
coldata %>%
  ggplot(aes(x = donor, fill = group, color = Sample_Name)) +
  geom_bar() +
  labs(x = "Donor", fill = "Group", color = "Sample Name")
```

# Pre-processing Workflow

## Filtering

```{r, message=FALSE}
data_t <- data[, -1] %>% t()
rownames(data_t) <- colnames(data[, -1])
colnames(data_t) <- data$Cell_Index

sce <- SingleCellExperiment(assays = list(counts = data_t), colData = coldata)
rowData(sce)$Type <- ifelse(grepl("abo", rownames(sce), ignore.case = TRUE), "Antibody Capture", "Gene Expression") # 'Type' is either gene or antibody
```

```{r, message=FALSE}
# removing the cells with multiplet or undetermined assignment
sce <- sce[, !coldata$invalid]

coldata <- coldata %>%
  dplyr::filter(!invalid)
```

```{r, message=FALSE}
# splitting protein from RNA
sce_split <- splitAltExps(sce, rowData(sce)$Type)
# altExpNames(sce_split)
```

### Overall Quality Metrics

#### Before Filtering

```{r, message=FALSE}
# ADT vs RNA counts
rna_numi <- colSums(counts(sce_split))
adt_numi <- colSums(counts(altExp(sce_split)))
nUMIdf <- data.frame(rna_nUMI = rna_numi,
                     adt_nUMI = adt_numi)

ggplot(nUMIdf, aes(x = log10(adt_nUMI+1), y = log10(rna_nUMI+1))) + 
  geom_point(alpha = 0.1, size = 1) +
  geom_density_2d(color = "orange") +
  labs(x = "ADT Counts", y = "RNA Counts")
```

#### Computing QC metrics

```{r scDblFinder, echo=TRUE, include=FALSE}
# detecting doublets using scDblFinder
# keeping the doublets returned and rechecking them in downstream analysis
set.seed(42L)

sce_split <- scDblFinder(sce_split)
```

```{r, message=FALSE}
outliers <- perCellQCMetrics(sce_split)
libsize.drop <- isOutlier(outliers$sum, log = TRUE, type = "both", nmads = 3)
feature.drop <- isOutlier(outliers$detected, log = TRUE, type = "lower", nmads = 3)
print(data.frame(ByLibSize = sum(libsize.drop),
                 ByFeature = sum(feature.drop),
                 Discard = sum(libsize.drop | feature.drop)))
outliers.discard <- libsize.drop | feature.drop

#outliers.discard <- quickPerCellQC(outliers)
#colSums(as.matrix(outliers.discard))
```

#### Ab Counts

```{r, message=FALSE}
ab.discard <- isOutlier(outliers$`altexps_Antibody Capture_detected`,
                        log = TRUE,
                        type = "lower",
                        min_diff = 1)
summary(ab.discard)

hist(outliers$`altexps_Antibody Capture_detected`, 
     col = 'grey',
     main = "", xlab = "Number of detected ADTs")
```

#### QC Plots

```{r, message=FALSE}
colData(sce_split) <- cbind(colData(sce_split), outliers) # adding the outliers info (sum, detected) to the sce object coldata
sce_split$discard <- outliers.discard | ab.discard # adding the cells to be discarded to the sce object coldata

plotColData(sce_split, x = "detected", y = "sum", colour_by = "discard") +
  theme(panel.border = element_rect(color = "grey")) + 
  labs(x = "Number of Features Detected", y = "Library Size")
```

```{r, message=FALSE}
gridExtra::grid.arrange(
  plotColData(sce_split, y="sum", colour_by="discard") +
    scale_y_log10() + ggtitle("Library Size"),
  plotColData(sce_split, y="detected", colour_by="discard") +
    scale_y_log10() + ggtitle("Genes Detected") + labs(y = "genes detected"),
  plotColData(sce_split, y="altexps_Antibody Capture_detected", colour_by="discard") +
    scale_y_log10() + ggtitle("Antibodies Detected") + labs(y = "antibodies detected"),
  ncol=1
)
```

```{r, message=FALSE}
# ADT vs RNA counts
rna_numi <- colSums(counts(sce_split))
adt_numi <- colSums(counts(altExp(sce_split)))
nUMIdf <- data.frame(rna_nUMI = rna_numi,
                     adt_nUMI = adt_numi, 
                     Discard = sce_split$discard)

ggplot(nUMIdf, aes(x = log10(adt_nUMI+1), y = log10(rna_nUMI+1), color = Discard)) + 
  geom_point(alpha = 0.1, size = 1) +
  geom_density_2d(color = "black") +
  labs(x = "ADT Counts", y = "RNA Counts") +
  scale_color_manual(values = c("#0072B2", "#C4961A"))
```

```{r filtering, message=FALSE}
# filtering; removing outliers
sce_split <- sce_split[, !sce_split$discard]
sce_split
table(coldata$Sample_Name)
```

## Normalization

```{r, message=FALSE}
# RNA

## size factors
lib.sf <- librarySizeFactors(sce_split)
summary(lib.sf)

lib.sf.df <- data.frame(lib.sf, Group = colData(sce_split)$group)

## plotting size factors
hist(log10(lib.sf), xlab="Log10[Size Factor]", col='grey80')

ggplot(lib.sf.df, aes(log10(lib.sf))) +  
  geom_histogram() + 
  facet_wrap(~Group, scales = "free") + 
  labs(x = "Log10[Size Factor]")

## normalization
sce_split <- logNormCounts(sce_split)
#assayNames(sce_split)

# ADT

clr_norm <- function(x) {
  return(log1p(x = x / (exp(x = sum(log1p(x = x[x > 0]), na.rm = TRUE) / length(x = x)))))
}

logcounts(altExp(sce_split)) <- apply(counts(altExp(sce_split)), 2, clr_norm)
```

## Variance Modeling

```{r, message=FALSE}
# determining variance components
dec <- modelGeneVar(sce_split)
dec[order(dec$bio, decreasing = TRUE), ]

fit <- metadata(dec)
plot(fit$mean, fit$var, 
     xlab = "Mean of log-expression", ylab = "Variance of log-expression")
points(fit$mean, fit$var, col = "red", pch = 16)
curve(fit$trend(x), col = "dodgerblue", add = TRUE, lwd = 2)
```

## Distribution of ADT Markers 

Looking at the distribution of ADT markers (after normalization): 

```{r, message=FALSE, fig.asp=1.8}
p <- do.call(plot_grid, c(lapply(rownames(logcounts(altExp(sce_split))), function(adt_name){
  ggplot(as.data.frame(logcounts(altExp(sce_split))[adt_name, ]), aes_string(x=logcounts(altExp(sce_split))[adt_name, ])) +
    geom_histogram(color="black",
                   fill = "black",
                   breaks=seq(0, 4.5, by=0.10)) + 
    xlab(adt_name) + 
    theme(axis.title.x = element_text(size = 8, face="bold"))}), ncol = 4))

print(p)
```

We define a new column containing the condition of samples:

- NT = NTbefore + NTafter
- KO = KObefore + KOafter
- CAR = CARbefore + CARafter
- CARKO = CARKObefore + CARKOafter

```{r, message=FALSE}
# creating new colData
sce_split$CAR <- ifelse(sce_split$Sample_Name %in% c("CARKOafter", "CARKObefore", "CARafter", "CARbefore"), "Yes", "No")
sce_split$KO <- ifelse(sce_split$Sample_Name %in% c("CARKOafter", "CARKObefore", "KOafter", "KObefore"), "Yes", "No")
sce_split$condition <- ifelse(sce_split$Sample_Name %in% c("CARKOafter", "CARKObefore"), "CARKO",
                                 ifelse(sce_split$Sample_Name %in% c("CARafter", "CARbefore"), "CAR",
                                        ifelse(sce_split$Sample_Name %in% c("KOafter", "KObefore"), "KO", "NT")))
sce_split$condition <- factor(sce_split$condition, levels = c("NT", "KO", "CAR", "CARKO"))
```

```{r}
saveRDS(sce_split, "intermediate_files/D2/sce_donor2.Rds")
```

# Session

```{r sessionInfo, cache = 0}
date()
sessionInfo()

knitr::knit_exit()
```



